celements-model/src/main/java/com/celements/store/part/DocumentSavePreparationCommand.java

```diff
@@ -38,57 +37,130 @@
+        doc.setElement(XWikiDocument.HAS_ATTACHMENTS, false);
       }
-    } else {
-      doc.setElement(XWikiDocument.HAS_OBJECTS, false);
-      doc.setElement(XWikiDocument.HAS_ATTACHMENTS, false);
+    } catch (IdComputationException exc) {
+      throw new XWikiException(MODULE_XWIKI_STORE, ERROR_XWIKI_STORE_HIBERNATE_SAVING_DOC,
+          "saveXWikiDoc - [" + getDatabase() + "] failed to compute id for [" + doc + "]", exc);
     }
+    return this;
+  }
+
+  private void prepareSession(boolean bTransaction) throws XWikiException {
+    checkState(session == null, "command already executed");
     if (bTransaction) {
       store.checkHibernate(context);
       SessionFactory sfactory = store.injectCustomMappingsInSessionFactory(doc, context);
       store.beginTransaction(sfactory, context);
     }
-    Session session = store.getSession(context);
+    session = store.getSession(context);
     session.setFlushMode(FlushMode.COMMIT);
-    return session;
   }
 
-  private void ensureDatabaseConsistency(XWikiDocument doc, XWikiContext context) {
-    WikiReference wikiRef = new WikiReference(context.getDatabase());
-    if (!wikiRef.equals(doc.getDocumentReference().getWikiReference())) {
-      LOGGER.warn("saveXWikiDoc - not matching wiki, adjusting to [{}] for doc [{}]", wikiRef,
-          doc.getDocumentReference(), new Throwable());
+  private void ensureDatabaseConsistency() {
+    if (!doc.getDocumentReference().getWikiReference().getName().equals(getDatabase())) {
+      LOGGER.warn("saveXWikiDoc - [{}] not matching database, adjusting from doc [{}]",
+          getDatabase(), doc.getDocumentReference(), new Throwable());
       boolean isMetaDataDirty = doc.isMetaDataDirty();
-      doc.setDocumentReference(References.adjustRef(doc.getDocumentReference(),
-          DocumentReference.class, wikiRef));
+      doc.setDocumentReference(RefBuilder.from(doc.getDocumentReference()).wiki(getDatabase())
+          .build(DocumentReference.class));
       doc.setMetaDataDirty(isMetaDataDirty);
     }
   }
 
+  private void prepareDocId() throws IdComputationException, HibernateException {
+    checkState(hasExistingDoc == null, "command already executed");
+    if (doc.hasValidId()) {
+      // documents with valid id have been loaded from the store, thus no id computation needed
+      hasExistingDoc = true;
+    } else {
+      doc.setId(computeNextFreeDocId(), store.getIdComputer().getIdVersion());
+    }
+  }
+
+  private long computeNextFreeDocId() throws IdComputationException, HibernateException {
+    String fullName = doc.getFullName();
+    long docId;
+    String existingFullName;
+    byte collisionCount = -1;
+    do {
+      collisionCount++;
+      docId = store.getIdComputer().computeDocumentId(
+          doc.getDocumentReference(), doc.getLanguage(), collisionCount);
+      existingFullName = loadExistingDocForId(docId);
+    } while ((existingFullName != null) && hasCollision(fullName, existingFullName, docId));
```

Comment:
as discussed: please take language into comparison

Reply:
- language has to be considered for collision calculations
- load all ids with lang in a single query (to be used in load and save), no caching
- fast fail if existing document is the same as to be loaded (fn&lang)
- address distributed concurrency?